### ISSUE TITLE
(first-order) Implication constraints

### ISSUE THREAD

**Issue by phadej** [haskell/cabal#8035]
```
constraints: basement ==0.0.13 => base >=4.10
```

That syntax might be hard to parse, and I'd be happy with any variant.

There is a practical need for this: some maintainers have explicitly forbidden trustees to make revisions to their packages, but we still need to check whether other stuff builds. It would help if `constraints` would allow to guide solver as if the revisions where in package itself. Revisions are still more powerful, but I think (first-order) implication constraint may help in many situations and shouldn't be too hard to implement.

**Comment by gbaz** [comment]
Is this proposed for just cabal.project files, or also build-depends fields?

**Comment by phadej** [comment]
Just `cabal.project` files.

Implication constraints will be difficult to implement for installed packages, e.g. if `text` had something like `bytestring <0.11 => ...`, as installed package info is recorded in `InstalledPackageInfo` which is part of `ghc-pkg`, so cannot be changed for older GHCs. That's a similar problem as with  with public sublibraries, challenge of making installed libraries work properly. Which is a real issue e.g. in nixpkgs, where everything is provided in global package db.

**Comment by emilypi** [comment]
I vote no until we have an actual grammar for cabal.project files specified, and then figure out how this should work. We should not be embedding syntax into the build tool just because some maintainers are assholes and we need to work around them. 

There is already a remedy for the problem statement: fork the package or use a source repository to pin an updated version. If a fork is warranted, urge downstream users to migrate because the fork.

**Comment by andreabedini** [comment]
I don't understand the goal. If this is only in a cabal.project, one can unpack the desired version of a package in the project folder and make the revision there. 

FIWI I agree with @emilypi (minus the moral judgment on the maintainers). There are many ways around this problem, we don't need to overcomplicate our tools.

**Comment by phadej** [comment]
> one can unpack the desired version of a package in the project folder and make the revision there.

that doesn't work for trustee workflow. if package is local, it have to be built. my need with implication constraints is to tell solver that **if** it going to pick a specific package version it should satisfy extra constraints. Not selecting that particular package version should be ok.

**Comment by andreabedini** [comment]
@phadej thank you for breaking this down for me, I admit I don't know the trustee workflow.

> my need with implication constraints is to tell solver that if it going to pick a specific package version it should satisfy extra constraints.

If I can a bit more of your patience, may I ask in what cases is this required?

**Comment by phadej** [comment]
@andreabedini i explain the use case in the issue description.

**Comment by phadej** [comment]
I could use `file+local` repositories overriding the Hackage version. I'll do that. thanks.
